src/create_torrent.cpp

```diff
@@ -518,38 +507,94 @@ namespace {
 				add_http_seed(s.url);
 		}
 
-		if (!m_v2_only)
+		if (make_v1)
 		{
 			m_piece_hash.resize(m_files.num_pieces());
 			for (auto const i : m_files.piece_range())
 				set_hash(i, ti.hash_for_piece(i));
 		}
 
-		if (!m_v1_only)
+		if (make_v2)
 		{
+//			auto const& file_trees = ti.internal_merkle_trees();
 			m_fileroots.resize(m_files.num_files());
 			m_file_piece_hash.resize(m_files.num_files());
 			for (auto const i : m_files.file_range())
 			{
 				// don't include merkle hash trees for pad files
 				if (m_files.pad_file_at(i)) continue;
 				m_file_piece_hash[i].resize(std::size_t(m_files.file_num_pieces(i)));
+
+				// TODO: copy piece layer from file_trees[i]
 			}
 		}
 
 		auto const info = ti.info_section();
 		m_info_dict.preformatted().assign(info.data(), info.data() + info.size());
 	}
 
-	entry create_torrent::generate() const
+namespace {
+	bool validate_v2_hashes(file_storage const& fs
+		, aux::vector<aux::vector<sha256_hash, piece_index_t::diff_type>, file_index_t> const& file_piece_hash)
 	{
-		entry dict;
+		if (int(file_piece_hash.size()) != fs.num_files()) return false;
+
+		int const piece_size = fs.piece_length();
 
+		for (auto i : fs.file_range())
+		{
+			auto const& hashes = file_piece_hash[i];
+
+			// pad files are not supposed to have any hashes
+			if (fs.pad_file_at(i))
+			{
+				if (!hashes.empty()) return false;
+				continue;
+			}
+
+			if (int(hashes.size()) != (fs.file_size(i) + piece_size - 1) / piece_size) return false;
+			if (std::any_of(hashes.begin(), hashes.end(), [](sha256_hash const& h)
+				{ return h.is_all_zeros(); }))
+			{
+				return false;
+			}
+		}
+		return true;
+	}
+
+	bool validate_v1_hashes(file_storage const& fs
+		, aux::vector<sha1_hash, piece_index_t> piece_hash)
+	{
+		int const piece_size = fs.piece_length();
+		if (int(piece_hash.size()) != (fs.total_size() + piece_size - 1) / piece_size)
+			return false;
+
+		return !std::any_of(piece_hash.begin(), piece_hash.end()
+			, [](sha1_hash const& h) { return h.is_all_zeros(); });
+	}
+}
+
+	entry create_torrent::generate() const
+	{
 		if (m_files.num_files() == 0 || m_files.total_size() == 0)
-			return dict;
+			aux::throw_ex<system_error>(errors::torrent_missing_file_tree);
+
+		// if all v2 hashes are set correctly, generate the v2 parts of the
+		// torrent
+		bool const make_v2 = validate_v2_hashes(m_files, m_file_piece_hash);
+		bool const make_v1 = validate_v1_hashes(m_files, m_piece_hash);
+
+		// if neither v1 nor v2 hashes were set, we can't create a torrent
+		if (!make_v1 && !make_v2)
+			aux::throw_ex<system_error>(errors::invalid_hash_entry);
+
+		if ((make_v1 && m_v2_only) || (make_v2 && m_v1_only))
+			aux::throw_ex<system_error>(errors::invalid_hash_entry);
```

Comment:
maybe this check is excessive. or maybe it would make more sense to have `set_hash()` and `set_hash2()` fail directly instead of deferring the failure until here.

Reply:
I think it makes more sense to do this check in `set_hash()` and `set_hash2()`.